File: homework-app/src/mockApi.go
package mockapi

import (
	"fmt"
	"time"
)

// TaskStatus は宿題のステータスを表す独自型
type TaskStatus string

const (
	StatusPending   TaskStatus = "pending"
	StatusSubmitted TaskStatus = "submitted"
	StatusApproved  TaskStatus = "approved"
)

// Task は宿題の型定義
type Task struct {
	ID      string     `json:"id"`
	Title   string     `json:"title"`
	DueDate string     `json:"dueDate"`
	Status  TaskStatus `json:"status"`
}

// File はブラウザの File オブジェクトの代わりとなるモック構造体
type File struct {
	Name string
	Size int64
}

// 1. OTPリクエスト（ダミー）
func RequestOTP(email string) (bool, error) {
	fmt.Printf("[API Mock] %s 宛にOTPを送信しました。ダミーOTPは「123456」です。\n", email)
	time.Sleep(1 * time.Second)

	return true, nil
}

// 2. OTP検証（ダミー）
func VerifyOTP(email string, otp string) (string, error) {
	time.Sleep(1 * time.Second)

	if otp == "123456" {
		return "dummy-jwt-token", nil
	}

	// 不正なOTPの場合は空文字（またはエラー）を返す
	return "", fmt.Errorf("無効なOTPです")
}

// 3. 宿題一覧取得（ダミー）
func FetchTasks() ([]Task, error) {
	time.Sleep(500 * time.Millisecond)

	tasks := []Task{
		{ID: "1", Title: "青チャート P.50〜55", DueDate: "2026-05-20", Status: StatusPending},
		{ID: "2", Title: "英単語ターゲット 1-100", DueDate: "2026-05-21", Status: StatusPending},
		{ID: "3", Title: "物理のエッセンス 力学", DueDate: "2026-05-18", Status: StatusSubmitted},
	}

	return tasks, nil
}

// 4. 画像アップロード＆完了報告（ダミー）
func SubmitTask(taskID string, file File) (bool, error) {
	fmt.Printf("[API Mock] タスク %s に画像 %s をアップロード（Signed URL経由を想定）\n", taskID, file.Name)
	time.Sleep(1500 * time.Millisecond)

	return true, nil
}

File: homework-app/src/mockApi.ts
// src/mockApi.ts
// src/api.ts
// ※フロントエンドからバックエンド（サーバー）へリクエストを送るためのファイルです

declare var process: {
  env: {
    NEXT_PUBLIC_API_URL?: string;
    [key: string]: string | undefined;
  };
};

export type TaskStatus = 'pending' | 'submitted' | 'approved';
export interface Task {
  id: string;
  title: string;
  dueDate: string;
  status: TaskStatus;
}

// 本番環境のバックエンドURLを環境変数から取得（なければローカル開発用）
// 実行時に process がない環境での ReferenceError を防ぐ
const API_BASE_URL = (typeof process !== 'undefined' && process.env.NEXT_PUBLIC_API_URL) 
  ? process.env.NEXT_PUBLIC_API_URL 
  : 'http://localhost:8000/api';

// 1. OTPリクエスト（本番用）
// ...existing code...

// 1. OTPリクエスト（本番用）
export async function requestOtp(email: string): Promise<boolean> {
  const response = await fetch(`${API_BASE_URL}/auth/request-otp`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ email }),
  });
  
  if (!response.ok) {
    throw new Error('OTPの送信に失敗しました');
  }
  return true;
}

// 2. OTP検証（本番用）
export async function verifyOtp(email: string, otp: string): Promise<string | null> {
  const response = await fetch(`${API_BASE_URL}/auth/verify-otp`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ email, otp }),
  });
  
  if (!response.ok) {
    return null; // 認証失敗時はnullを返す
  }
  
  const data = await response.json();
  return data.token; // バックエンドが発行したJWTなどを返す
}

// 3. 宿題一覧取得（本番用）
// ※誰の宿題か特定するために、ログイン時に取得したtokenを受け取るように変更
export async function fetchTasks(token: string): Promise<Task[]> {
  const response = await fetch(`${API_BASE_URL}/tasks`, {
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
  });
  
  if (!response.ok) {
    throw new Error('タスクの取得に失敗しました');
  }
  
  return await response.json();
}

// 4. 画像アップロード＆完了報告（本番用）
// ※画像ファイルはJSONに含められないため、FormDataを使用します
export async function submitTask(taskId: string, file: File, token: string): Promise<boolean> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch(`${API_BASE_URL}/tasks/${taskId}/submit`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      // 注意: FormDataを送る際は 'Content-Type' を手動で指定しません（ブラウザが自動で設定します）
    },
    body: formData,
  });

  if (!response.ok) {
    throw new Error('画像のアップロードに失敗しました');
  }
  return true;
}